Add first-class Little Endian support for Int to ByteBuf and descendants

ifesdjeen opened the pull request:
As discussed in #3209, this PR adds Little Endian accessors 
to ByteBuf and descendants.

Corresponding accessors were added to UnsafeByteBufUtil,
HeapByteBufferUtil to avoid calling `Integer.reverseBytes`.

ifesdjeen commented:
This PR is work-in-process, just wanted to discuss the API with @normanmaurer. I've added `getLEInt` and `setLEInt` so far, although it won't be a problem to add similar accessors for all the `ByteBuf` instances. 

I thought that adding a method with `ByteOrder` and `if` might be costly, so I went with @trustin's suggestion and just added `LE` accessors along with the "usual" ones. 

If you're fine with names, I'll add more tests and prepare same for the rest of "affected" data types.

ifesdjeen commented:
So far, `int` and `short` are implemented for random access, should we add Sequential Access accessors, too @normanmaurer?

trustin commented:
We should also have sequential accessors. I wonder if `get<TYPE>LE` is better (or not.) Opinions?

normanmaurer commented:
@trustin @ifesdjeen I agree I kind of like `get<TYPE>LE` more from the style. Just not sure if it may make it easier for people to use the "wrong method". The other option which I still like is `get<TYPE>(int, ByteOrder)`.

ifesdjeen commented:
changed the signatures to `get<TYPE>LE`.

The only reason I didn't do `get<TYPE>(int, ByteOrder)` is because it'd involve if/else statements, which'd take more time to process (not sure if they'd be optimised out eventually). 

Other the other hand, that'd allow things like switching the endianness by just changing it all in one place.

I'll move to `unsigned` and sequential for now, if in the end it's decided to still go with the second option, I'll make the change.

ifesdjeen commented:
Ok, I've added sequential and unsigned for all the datatypes in charge, added tests for sequential, random access and for after-release.

It's ready for your review @normanmaurer @trustin

ifesdjeen commented:
Could you take a short look? 

It's all covered with tests, API is adjusted as Trustin said, most likely the only two files really worth looking at are `UnsafeByteBufUtil` and `HeapByteBufferUtil`, as there are some bit manipulation, other than that it's mostly calling internal APIs.

normanmaurer commented:
@ifesdjeen awesome! Looks good..

@trustin PTAL as well.

Scottmitch commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -648,6 +660,20 @@ protected int _getUnsignedMedium(int index) {
+        if (index + 3 <= c.endOffset) {
+            return c.buf.getUnsignedMediumLE(index - c.offset);
+        } else if (order() == ByteOrder.BIG_ENDIAN) {
+            return _getShortLE(index) & 0xFFFF |
```
nit: consider putting this on 1 line and also consider using lowercase `f`s for consistency.

ifesdjeen replied:
oh, TBH, I used the same style as the `_getUnsignedMedium`, I can change both, just give me a sec.

ifesdjeen replied:
on the second note, I've found a bunch of places where it's mixed capitaized/lowercase. I will fix it here for now, can also submit a PR with a cleanup everyehere if you think it's a good idea..

Scottmitch replied:
@ifesdjeen - Yip I realized you were just following what was already there...but might as well make more consistent while we are here.  Thanks :smile:

Scottmitch commented on buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java:
```diff
@@ -648,6 +660,20 @@ protected int _getUnsignedMedium(int index) {
+            return _getShortLE(index) & 0xFFFF |
+                   (_getByte(index + 2) & 0xFF) << 16;
+        } else {
+            return (_getShortLE(index) & 0xffff) << 8 |
```
nit: consider putting on 1 line. This will apply to all lines in this file (which don't exceed the 120 character limit).

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/DuplicatedByteBuf.java:
```diff
@@ -146,6 +156,11 @@ protected int _getInt(int index) {
 
     @Override
+    protected int _getIntLE(int index) {
+        return ByteBufUtil.swapInt(buffer.getInt(index));
```
`buffer.getIntLE(index);`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/DuplicatedByteBuf.java:
```diff
@@ -228,6 +258,11 @@ protected void _setInt(int index, int value) {
 
     @Override
+    protected void _setIntLE(int index, int value) {
+        buffer.setInt(index, ByteBufUtil.swapInt(value));
```
`buffer.setIntLE(index, value);`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/FixedCompositeByteBuf.java:
```diff
@@ -302,6 +358,20 @@ protected long _getLong(int index) {
+        if (index + 8 <= c.endOffset) {
+            return c.buf.getLongLE(index - c.offset);
+        } else if (order() == ByteOrder.BIG_ENDIAN) {
+            return _getIntLE(index) & 0xFFFFFFFFL |
```
nit: keep on the same line.

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/PooledDirectByteBuf.java:
```diff
@@ -66,9 +66,24 @@ protected short _getShort(int index) {
 
     @Override
+    protected short _getShortLE(int index) {
+        return ByteBufUtil.swapShort(memory.getShort(idx(index)));
```
can we just use `memory.getShortLE(idx(index));`?

Same question for other methods which use `ByteBufUtil` in this class.

Scottmitch replied:
consider just calling `return ByteBufUtil.swapShort(_getShort(index));`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/PooledDirectByteBuf.java:
```diff
@@ -66,9 +66,24 @@ protected short _getShort(int index) {
+    @Override
+    protected int _getUnsignedMediumLE(int index) {
+        index = idx(index);
+        return memory.get(index + 2)  & 0xff        |
```
this method looks equivalent to `_getUnsignedMedium`

ifesdjeen replied:
True, it should've been

```java
   @Override
    protected int _getUnsignedMediumLE(int index) {
        index = idx(index);
        return memory.get(index)  & 0xff        |
               (memory.get(index + 1) & 0xff) << 8  |
               (memory.get(index + 2) & 0xff) << 16 ;
    }
```

Scottmitch commented on buffer/src/main/java/io/netty/buffer/PooledUnsafeDirectByteBuf.java:
```diff
@@ -85,21 +85,41 @@ protected short _getShort(int index) {
 
     @Override
+    protected short _getShortLE(int index) {
+        return UnsafeByteBufUtil.getLEShort(addr(index));
```
`getLEShort` -> `getShortLE`.

rename methods in `UnsafeByteBufUtil` to be consistent with the `ByteBuf` methods.

ifesdjeen replied:
fixed all...

Scottmitch commented on buffer/src/main/java/io/netty/buffer/SlicedByteBuf.java:
```diff
@@ -125,21 +125,41 @@ protected short _getShort(int index) {
 
     @Override
+    protected int _getIntLE(int index) {
+        return ByteBufUtil.swapInt(buffer.getInt(idx(index)));
```
consider just using `ByteBufUtil.swapInt(_getInt(index));`

Scottmitch replied:
`return buffer.getIntLE(idx(index));`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/SlicedByteBuf.java:
```diff
@@ -190,21 +210,41 @@ protected void _setShort(int index, int value) {
 
     @Override
+    protected void _setIntLE(int index, int value) {
+        buffer.setInt(idx(index), ByteBufUtil.swapInt(value));
```
`buffer.setIntLE(idx(index), value);`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/UnpooledUnsafeDirectByteBuf.java:
```diff
@@ -224,21 +224,41 @@ protected short _getShort(int index) {
 
     @Override
+    protected short _getShortLE(int index) {
+        return UnsafeByteBufUtil.getLEShort(addr(index));
```
rename new methods in `UnsafeByteBufUtil` to be consistent with `ByteBuf` naming scheme.

`getLEShort` -> `getShortLE`.

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java:
```diff
@@ -193,6 +326,21 @@ static long getLong(byte[] array, int index) {
+    static long getLELong(byte[] array, int index) {
+        if (UNALIGNED) {
+            long v =  PlatformDependent.getLong(array, index);
+            return BIG_ENDIAN_NATIVE_ORDER ? v : Long.reverseBytes(v);
```
`return BIG_ENDIAN_NATIVE_ORDER ? Long.reverseBytes(v) : v`

ifesdjeen replied:
oh my... this one is painful(
sorry. And Fixed...

Scottmitch replied:
np ... endianness related changes (and reviews) are usually painful :cry:

Scottmitch commented on buffer/src/main/java/io/netty/buffer/UnsafeByteBufUtil.java:
```diff
@@ -193,6 +326,21 @@ static long getLong(byte[] array, int index) {
 
+    static long getLELong(byte[] array, int index) {
+        if (UNALIGNED) {
+            long v =  PlatformDependent.getLong(array, index);
```
kill double space after `=  `.

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/WrappedByteBuf.java:
```diff
@@ -224,36 +224,71 @@ public short getShort(int index) {
 
     @Override
+    public short getShortLE(int index) {
+        return ByteBufUtil.swapShort(buf.getShort(index));
```
`return buf.getShortLE(index);`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/WrappedByteBuf.java:
```diff
@@ -224,36 +224,71 @@ public short getShort(int index) {
 
     @Override
+    public int getIntLE(int index) {
+        return ByteBufUtil.swapInt(buf.getInt(index));
```
`buf.getIntLE(index);`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/main/java/io/netty/buffer/WrappedByteBuf.java:
```diff
@@ -334,24 +369,48 @@ public ByteBuf setShort(int index, int value) {
 
     @Override
+    public ByteBuf setShortLE(int index, int value) {
+        buf.setShort(index, ByteBufUtil.swapShort((short) value));
```
`buf.setShortLE(index, value);`

ifesdjeen replied:
fixed

normanmaurer commented:
I will have a detailed look tomorrow morning

Scottmitch commented on buffer/src/main/java/io/netty/buffer/WrappedByteBuf.java:
```diff
@@ -334,24 +369,48 @@ public ByteBuf setShort(int index, int value) {
 
     @Override
+    public ByteBuf setIntLE(int index, int value) {
+        buf.setInt(index, ByteBufUtil.swapInt(value));
```
`buf.setIntLE(index, value);`

ifesdjeen replied:
fixed

Scottmitch commented on buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java:
```diff
@@ -593,6 +719,34 @@ public void testSequentialShortAccess() {
+            short value = (short) random.nextInt();
+            assertEquals(i, buffer.writerIndex());
+            assertTrue(buffer.isWritable());
+            buffer.writeShortLE(value);
```
consider sharing code with the previous test. This line looks like the only difference?  Maybe just have a method:

```java
private static void testSequentialShortAccees(boolean testBigEndian) {
..
}
```

Scottmitch commented on buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java:
```diff
@@ -621,6 +775,34 @@ public void testSequentialUnsignedShortAccess() {
+            short value = (short) random.nextInt();
+            assertEquals(i, buffer.writerIndex());
+            assertTrue(buffer.isWritable());
+            buffer.writeShortLE(value);
```
can we share code with `testSequentialUnsignedShortAccess`?

Scottmitch replied:
This would be a general comment for all new methods in this file.

ifesdjeen replied:
yup, doing that already

Scottmitch commented on buffer/src/test/java/io/netty/buffer/AbstractByteBufTest.java:
```diff
@@ -1997,16 +2319,31 @@ public void testGetShortAfterRelease() {
 
     @Test(expected = IllegalReferenceCountException.class)
+    public void testGetMediumLEAfterRelease() {
+    releasedBuffer().getMediumLE(0);
```
indentation?

ifesdjeen replied:
fixed

Scottmitch commented on codec/src/main/java/io/netty/handler/codec/ReplayingDecoderByteBuf.java:
```diff
@@ -246,42 +247,84 @@ public int getInt(int index) {
     @Override
+    public int getIntLE(int index) {
+        checkIndex(index, 4);
+        return ByteBufUtil.swapInt(buffer.getInt(index));
```
`return buffer.getIntLE(index);`

ifesdjeen replied:
fixed

Scottmitch commented:
@ifesdjeen - great work! I finished a first pass. ping me when ready for another round of review.

ifesdjeen commented:
@Scottmitch sorry about so much stupidity from my side... after you pointed up first few things, I got very ashamed :(

Scottmitch commented:
@ifesdjeen - Don't be ashamed at all! Code is generally of good quality and that is what reviews are for.  Keep up the good work!

ifesdjeen commented:
I've found a coule more small things after @Scottmitch pointers.

At the moment the only thing I could find was medium formatting: https://github.com/netty/netty/pull/4477/files#diff-bc484e7bb4f602319a15eb2b5e9d7a5fL83 
although it overflows 120 symbols plus when it's 3 segments, it's much easier to read it that way.

I promise I'll be more rigorous with PR next time...

normanmaurer commented on buffer/src/main/java/io/netty/buffer/ByteBuf.java:
```diff
@@ -1103,6 +1244,15 @@
 
     /**
+     * Gets an unsigned 16-bit short integer at the current {@code readerIndex}
+     * and increases the {@code readerIndex} by {@code 2} in this buffer.
```
Add to the javadocs that it uses little endian.

normanmaurer commented on buffer/src/main/java/io/netty/buffer/ByteBuf.java:
```diff
@@ -1121,6 +1281,16 @@
 
     /**
+     * Gets an unsigned 24-bit medium integer at the current {@code readerIndex}
+     * in the Little Endian Byte Order  and increases the {@code readerIndex}
```
remove extra whitespace after `Order`

ifesdjeen replied:
fixed

normanmaurer commented on buffer/src/main/java/io/netty/buffer/FixedCompositeByteBuf.java:
```diff
@@ -290,6 +334,18 @@ protected int _getInt(int index) {
+        } else {
+            return (_getShort(index) & 0xffff) << 16 | _getShort(index + 2) & 0xffff;
+        }
+    }
```
I think the above methods are shared between `FixedCompositeByteBuf` and `CompositeByteBuf`. Consider to create a CompositeByteBufUtils and move these there.

ifesdjeen replied:
Problem with extracting these to the common util class is that protected methods `_getShort` and so on are used. Also, that `Component` class is private (there are two of them, one is readonly and second one is read/write).

normanmaurer replied:
gotcha... ok leave it as it is for now.

ifesdjeen replied:
I could only come up with something like that:

```java
public class CompositeByteBufUtil {

    public static int getInt(ByteBuf buf, int index) {
        if (buf.order() == ByteOrder.BIG_ENDIAN) {
            return (buf.getShort(index) & 0xffff) << 16 | buf.getShort(index + 2) & 0xffff;
        } else {
            return buf.getShort(index) & 0xFFFF | (buf.getShort(index + 2) & 0xFFFF) << 16;
        }
    }

    public static int getIntLE(ByteBuf buf, int index) {
        if (buf.order() == ByteOrder.BIG_ENDIAN) {
            return buf.getShortLE(index) & 0xffff | (buf.getShortLE(index + 2) & 0xffff) << 16;
        } else {
            return (buf.getShortLE(index) & 0xffff) << 16 | buf.getShortLE(index + 2) & 0xffff;
        }
    }

}
```

(which would result into index check when calling next method).

If you think it's a good idea, I can make same change for all data types and push.
Otherwise - the rest of your comments are addressed :)

normanmaurer commented:
@ifesdjeen just two more small comments. @Scottmitch did all the heavy lifting already :) Thanks again to both of you guys!

normanmaurer commented:
@trustin would you mind having a look as well?

Scottmitch commented on buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java:
```diff
@@ -349,11 +349,24 @@ public short getShort(int index) {
+        return _getShortLE(index);
+    }
+
+    protected abstract short _getShortLE(int index);
```
I guess this change can only go to 4.1 @normanmaurer? If we wanted it to go back to 4.0 we could just use the regular getters and just use the swap utility methods.

normanmaurer replied:
4.1 only is fine

> Am 25.11.2015 um 22:39 schrieb Scott Mitchell <notifications@github.com>:
> 
> In buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java:
> 
> > @@ -349,11 +349,24 @@ public short getShort(int index) {
> >      protected abstract short _getShort(int index);
> >  
> >      @Override
> > +    public short getShortLE(int index) {
> > +        checkIndex(index, 2);
> > +        return _getShortLE(index);
> > +    }
> > +
> > +    protected abstract short _getShortLE(int index);
> I guess this change can only go to 4.1 @normanmaurer? If we wanted it to go back to 4.0 we could just use the regular getters and just use the swap utility methods.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Scottmitch commented on buffer/src/main/java/io/netty/buffer/ByteBuf.java:
```diff
@@ -558,6 +558,17 @@
+     *         if the specified {@code index} is less than {@code 0} or
+     *         {@code index + 2} is greater than {@code this.capacity}
+     */
+    public abstract short getShortLE(int index);
```
Same comment as https://github.com/netty/netty/pull/4477#discussion-diff-45922121. If we want this to go back into 4.0 we would have to make this non-abstract.

normanmaurer replied:
Only 4.1

> Am 25.11.2015 um 22:45 schrieb Scott Mitchell <notifications@github.com>:
> 
> In buffer/src/main/java/io/netty/buffer/ByteBuf.java:
> 
> > @@ -558,6 +558,17 @@
> >      public abstract short getShort(int index);
> >  
> >      /**
> > +     * Gets a 16-bit short integer at the specified absolute {@code index} in
> > +     * this buffer in Little Endian Byte Order. This method does not modify
> > +     * {@code readerIndex} or {@code writerIndex} of this buffer.
> > +     *
> > +     * @throws IndexOutOfBoundsException
> > +     *         if the specified {@code index} is less than {@code 0} or
> > +     *         {@code index + 2} is greater than {@code this.capacity}
> > +     */
> > +    public abstract short getShortLE(int index);
> Same comment as #4477 (diff). If we want this to go back into 4.0 we would have to make this non-abstract.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Scottmitch commented on buffer/src/main/java/io/netty/buffer/HeapByteBufUtil.java:
```diff
@@ -28,12 +28,22 @@ static short getShort(byte[] memory, int index) {
     }
 
+    static short getShortLE(byte[] memory, int index) {
+        return (short) (memory[index] & 0xFF | memory[index + 1] << 8);
```
`0xFF` -> `0xff`

Scottmitch commented on buffer/src/main/java/io/netty/buffer/UnpooledDirectByteBuf.java:
```diff
@@ -255,6 +269,11 @@ protected int _getInt(int index) {
 
     @Override
+    protected int _getIntLE(int index) {
+        return ByteBufUtil.swapInt(buffer.getInt());
```
`return ByteBufUtil.swapInt(_getInt(index));`

ifesdjeen replied:
Fixed

Scottmitch commented:
@ifesdjeen - Finished another pass. Looks good.

Just one question about how/if this will be back ported to 4.0. We can leave the abstract methods in 4.1, but in 4.0 I'm not sure if we can add abstract methods. We could just leave the PR as is for 4.1, and take care of this on the backport to 4.0 (if we plan on backporting). @normanmaurer - WDYT?

trustin commented:
2 things:

- Could we deprecate `order()`, `order(ByteBuf)` and `SwappedByteBuf`? A user doesn't really have to use `SwappedByteBuf` anymore.
- `CompositeByteBuf` could always use non-swapped buffer and then you can remove all the `order()` checks.

trustin commented:
I'd prefer moving forward with 4.1 fast and avoid backporting, but if someone is fine with maintenance, then I'm fine.

normanmaurer commented:
I think 4.1 only is fine here

> Am 26.11.2015 um 04:34 schrieb Trustin Lee <notifications@github.com>:
> 
> I'd prefer moving forward with 4.1 fast and avoid backporting, but if someone is fine with maintenance, then I'm fine.
> 
> —
> Reply to this email directly or view it on GitHub.
>

normanmaurer commented:
@ifesdjeen could you please address @trustin 's comments as well and ping me once done. Then I will merge :) Thanks again buddy!

ifesdjeen commented:
@normanmaurer sure sure :) currently doing that.

normanmaurer commented:
@ifesdjeen :beers: on me !

ifesdjeen commented:
I hope that's not the last patch I'm working on :P but :beers: with @normanmaurer always sound like a good idea.

ifesdjeen commented:
@normanmaurer I've deprecated `order()` and `order(buf)`, so far with annotations. I've split commits into two to separate and make the review simpler. I can squash them when they're reviewed.

There are still plenty checks for `order` across the code, maybe we should make the deprecated `order` and `order(buf)` throw an IllegalArgument exceptions? Questions is whether we'd like to leave the possibility to create a swapped buffer or no...

normanmaurer commented:
@ifesdjeen I think we need to revert the changes to CompositeByteBuf as this would be a breaking change. This rest looks good though

ifesdjeen commented:
Ah, I assumed the latter: 

 * `CompositeByteBuf` could always use non-swapped buffer and then you can remove all the order() checks.

Probably it was meant that in future this could be done. Ok, I have reverted the changes on `CompositeByteBuffer` to avoid assuming that it's all `BIG_ENDIAN`

normanmaurer commented:
Cherry-picked into 4.1 (0f9492c9affc528c766f9677952412564d4a3f6d) and master (920754ab0cfc9e91256739b54e0983222b4ef268). 

@ifesdjeen thanks!

ifesdjeen commented:
:fireworks: :dancer: :dancers:
Thank you a lot!
